device/transport/amqp/test/_amqp_receiver_test.js

```diff
@@ -41,153 +27,168 @@ describe('AmqpReceiver', function () {
+    it('removes the \'message\' and \'error\' listeners and detaches the underlying AMQP receiver if no one is listening', function() {
+      var recv = new AmqpReceiver(fakeConfig);
+      var fakeReceiverLink = new EventEmitter();
+      fakeReceiverLink.detach = sinon.stub().callsArg(0);
+      sinon.stub(fakeReceiverLink, 'removeListener');
+      recv._amqp = new FakeAmqp();
+      sinon.stub(recv._amqp, 'attachReceiverLink').callsArgWith(2, null, fakeReceiverLink);
       var fakeCallback = function() {};
       recv.on('message', fakeCallback);
       recv.removeListener('message', fakeCallback);
-      assert.isTrue(fakeAmqpReceiver.removeListener.calledWith('message'));
-    });
-
-    it('removes the errorReceived listener if no one is listening to message events', function() {
-      var fakeReceiver = new EventEmitter();
-      var fakeAmqpClient = {
-        getReceiver: function(ep, callback) {
-          callback(null, fakeReceiver);
-        }
-      };
-      var recv = new AmqpReceiver(fakeConfig, fakeAmqpClient, fakeMethodClient);
-      var fakeCallback1 = function() {};
-      var fakeCallback2 = function() {};
-      assert.strictEqual(fakeReceiver.listeners('errorReceived').length, 0);
-      recv.on('message', fakeCallback1);
-      assert.strictEqual(fakeReceiver.listeners('errorReceived').length, 1);
-      recv.on('message', fakeCallback2);
-      assert.strictEqual(fakeReceiver.listeners('errorReceived').length, 1);
-      recv.removeListener('message', fakeCallback1);
-      assert.strictEqual(fakeReceiver.listeners('errorReceived').length, 1);
-      recv.removeListener('message', fakeCallback2);
-      assert.strictEqual(fakeReceiver.listeners('errorReceived').length, 0);
+      assert.isTrue(fakeReceiverLink.removeListener.calledWith('message'));
+      assert.isTrue(fakeReceiverLink.removeListener.calledWith('error'));
+      assert.isTrue(fakeReceiverLink.detach.calledOnce);
     });
   });
 
-  describe('#on(<others>, callback)', function() {
-    it('does not forward event listeners other than message and errorReceived of the message event to the underlying AMQP receiver', function() {
-      var recv = new AmqpReceiver(fakeConfig, fakeAmqpClient, fakeMethodClient);
-      var fakeCallback = function() {};
-      recv.on('foo', fakeCallback);
-      assert(fakeAmqpReceiver.on.notCalled);
+  describe('#on(<others>, callback)', function () {
+    it('does not forward event listeners other than message and errorReceived of the message event to the underlying AMQP receiver', function () {
+      var recv = new AmqpReceiver(fakeConfig);
+      var fakeReceiverLink = new EventEmitter();
+      sinon.spy(fakeReceiverLink, 'on');
+      recv._amqp = new FakeAmqp();
+      sinon.stub(recv._amqp, 'attachReceiverLink').callsArgWith(2, null, fakeReceiverLink);
+      recv.on('foo', function () {});
+      assert(fakeReceiverLink.on.notCalled);
+      assert(recv._amqp.attachReceiverLink.notCalled);
```

Comment:
> assert(recv._amqp.attachReceiverLink.notCalled) [](start = 5, length = 48)

You've commented in the past about exposing the underlying implementation.  This (and a few other places) doesn't do that?

Reply:
yes and no. technically this is asserting something on the low boundary of the blackbox: making sure we're calling the right call. but at the same time we should be able to inject _amqp instead of reaching inside the private variable. so ultimately, you're right, but it's for a good purpose. we should refactor this nevertheless. not now though.

---
In reply to: [140305164](https://github.com/Azure/azure-iot-sdk-node/pull/87#discussion_r140305164) [](ancestors = 140305164)